[#31] 테마 및 디자인 게시판 관리 기능 추가

kym8821 opened the pull request:
## #️⃣연관된 이슈

> #31 

## 📝작업 내용
- 디자인 게시판 CRUD API 추가
- 테마 게시판 CRUD API 추가
- 디자인 게시판과 테마 게시판 기능에 대한 테스트 코드 작성
- 테스트 코드 공통 기능 분리 ( MockMvcUtils )
- 기타 빌드 에러 문제 해결

### 스크린샷 (선택)
- 빌드 및 테스트 성공 
<img width="375" height="54" alt="image" src="https://github.com/user-attachments/assets/ed028fe2-438c-40f0-a9d8-b9e0ad3f47b6" />  

- Design Board swagger 추가 확인
<img width="1453" height="354" alt="image" src="https://github.com/user-attachments/assets/b8a4e020-effb-41e4-b102-6d92805cecf0" />  

- Theme Board swagger 추가 확인
<img width="1459" height="344" alt="image" src="https://github.com/user-attachments/assets/6d2cf90f-3a51-4774-9ca6-30d6ed1724e9" />  


## 💬리뷰 요구사항(선택)

- 이해가 안되는 부분이나, 코드 작성 의도가 궁금하다면 바로 말씀해주세요
- 개선할 부분이 있다면, 바로 말씀해주세요

kym8821 committed:
feat : 페이지별 조회 시 페이지와 페이지 크기 기본값 설정 코드 개선

kym8821 committed:
fix : 게시글 수정 시 dto값 여부가 아닌 entity값 여부에 따라 값을 변경하던 문제 수정

kym8821 committed:
feat : 게시글 관련 Entity에서 양방향 다대일 관계를 단방향 다대일 관계로 수정
- 추후 양방향 다대일 관계가 필요하면 개선할 예정

kym8821 committed:
fix : 개발 환경의 가짜 데이터 생성 기능에서 사용자 이메일이 중복되는 문제 수정

kym8821 committed:
feat : 요청 DTO에 @Builder를 추가하여 테스트 환경에서 사용할 수 있도록 개선

kym8821 committed:
test : post 테스트 데이터 추가 기능과 MockMvc 유틸 기능 추가
- MockMvcUtils를 추가하여 인증 및 응답 객체 변환의 중복 코드 제거
- PostTestDataGenerator 코드를 추가

kym8821 committed:
test : 게시글과 댓글 테스트 코드 추가

kym8821 committed:
feat : dto projection 기반으로 post, 추천수, 태그 종류를 반환하는 기능 추가

kym8821 committed:
feat : PostService를 Post에 대한 생성, 조회, 수정, 삭제만을 갖는 서비스 클래스로 전환 후 facade 패턴 적용

kym8821 committed:
[merge] resolve conflict on config directory

kym8821 committed:
fix : 게시글 및 댓글 DTO를 API 명세서에 따라 수정

kym8821 committed:
refactor : 파사드 패턴을 적용하여 게시글 service 복합 기능 분리

kym8821 committed:
refactor : 파사드 패턴을 적용하여 댓글, 태그, 좋아요 service 복합 기능 분리

kym8821 committed:
test : 테스트 코드에 객체 이름 수정 내역 반영

kym8821 committed:
feat : 테마 게시판 기능 개발 및 테마 게시글 정보 테이블 분리
- ThemeBoard 테이블과 Post 테이블을 분리하여 게시글 공통 정보와 테마 게시글 정보 분리
- ThemeBoard 관련 DTO와 Post 관련 DTO 분리
- ThemeBoard 기반의 컨트롤러 개발

kym8821 committed:
feat : 개발용 테스트 데이터 개발 기능에 ThemeBoard와 ThemeComponent 추가

kym8821 committed:
test : MockMvcUtils에 테스트용 공통 기능 분리 및 ThemeBoard, Comment 테스트 코드 오류 개선

kym8821 committed:
fix : 조회 시 발생하는 themeComponent에 대한 N+1 문제 해결

kym8821 committed:
fix : ThemeComponent에 대한 fetch join 문법 오류 해결

kym8821 committed:
feat : [#31 - ]FileManager 인터페이스 행위 추가

FileManager 인터페이스에 이미지 경로 생성 행위 추가
S3FileManager의 resolveFilePath 접근자 Public로 변경

Fixes: #31

kym8821 committed:
feat : [#31 - ]QueryDSL 설정 추가

QueryDSL 의존성 추가
QueryDSL QClass 생성 위치 및 QClass 코드 인식 설정 추가
JpaQueryFactory 빈 추가

Fixes: #31

kym8821 committed:
feat: [#31 - ] PostEntity DTO 분리 및 프로필 이미지 정보 변경

이미지 이름을 기반으로 프로필 이미지 관리하도록 변경
ThemeBoard DTO와 Post DTO 분리

Fixes: #31

kym8821 committed:
feat: [#31 - ]PostSummary에 User 필드 추가 및 조회 기능 개선

PostRepositorySupport에서 QueryDSL 기반으로 PostSummary를 조회하도록 개선

Fixes: #31

kym8821 committed:
fix: [#31 - ] Json 필드를 스네이크 케이스로 변경

ThemeBoardDto, TagDto 내부 필드를 json 변환 시 스네이크 케이스로 인식하도록 개선

Fixes: #31

kym8821 committed:
feat: [#31 - ] 디자인 요소 게시판 Entity와 DTO 추가

DesignBoard : 디자인 요소 게시판 Entity
DesignBoardDto : 디자인 요소 게시판 Dto

Fixes: #31

kym8821 committed:
feat: [#31 - ] 게시판-좋아요 맵 조회 기능 추가

PreferRepositorySupport에 게시판-좋아요 수 매핑 기능 추가
Aggregate Count이기에 영속성 컨텍스트 캐싱 효과는 뭇받음

Fixes: #31

kym8821 committed:
feat: [#31 - ] 디자인 요소 도메인 서비스와 매퍼 추가

PostDtoMapper : 테마 및 디자인 요소 생성 / 수정 DTO를 게시판 생성/수정 DTO로 매핑함
DesignBoardService : 디자인 요소 게시글에 대한 데이터 유효성 검증 및 비즈니스 로직 수행

Fixes: #31

kym8821 committed:
feat: [#31 - ] 디자인 요소 게시판 리포지토리 레이어 추가

JpaRepository 기반 기본 제공 기능 추가
게시판 ID 기반 디자인 요소 게시판 조회/삭제 기능 추가

Fixes: #31

kym8821 committed:
feat: [#31 - ] 디자인 게시판에 대한 복합 도메인 기능 추가

DesignBoardManagementFacade : 디자인 게시글 복합 도메인 로직 담당
BoardManagementHelper : 게시글 로직에서 재사용 가능한 부분 담당 ( 프로필 이미지 관리 및 게시글 저장 등 )

Fixes: #31

kym8821 committed:
feat: [#31 - ] 디자인 요소 Entity 조회 기능 추가

DesignComponent에 디자인 요소 Entity 조회 기능 추가

Fixes: #31

kym8821 committed:
feat: [#31 - ] 테마 게시글 서비스 로직 개선

ThemeBoardManagementFacade의 게시글 작성자 유효성 검증 로직을 도메인 서비스의 책임으로 변경
게시글 생성 및 프로필 이미지 저장 기능을 BoardManagementHelper에서 재사용

Fixes: #31

kym8821 committed:
feat: [#31 - ] 테마 게시글 생성 시 이미지를 받도록 개선

POST /api/theme-boards : MultipartFile로 프로필 이미지를 받도록 개선

Fixes: #31

kym8821 committed:
chore: [#31 - ] 불필요한 코드 제거

ThemeBoardService에서 안쓰는 기능 제거 ( findAll )

Fixes: #31

kym8821 committed:
fix: [#31 - ] ThemeBoard 컨트롤러 파라미터 오류 수정

이미지와 DTO 받을 때 @RequestPart가 아닌 @RequestParam 사용하던 문제 수정

Fixes: #31

kym8821 committed:
feat: [#31 - ] ThemeBoard 플로우에서 불필요한 코드 제거 및 단일화

ThemeBoardDetailDto : DTO 변환 기능을 from 하나로 단일화
ThemeBoardManagementFacade : findThemeBoardPreviews에서 코드 간략화 및 불필요한 Map 전환 제거

Fixes: #31

kym8821 committed:
feat: [#31 - ] DesignBoard 컨트롤러 추가 및 불필요한 코드 제거

DesignBoardController : DesignBoard CRUD 컨트롤러 추가
DesignBoardManagementFacade : preview dto 조회 시 불필요한 매핑 코드 제거

Fixes: #31

kym8821 committed:
test: [#31 - ] form data 요청 테스트 코드 공통 기능 분리 및 코드 수정

MockMvcUtils : form data 생성 및 요청 코드 분리
ThemeBoardControllerTest : 게시판 생성 테스트 오류 수정

Fixes: #31

kym8821 committed:
fix: [#31 - ] 디자인 게시판 컨트롤러 문제 수정 및 불필요한 필드 제거

DesignBoardDto에서 불필요한 코드 제거
DesignBoardController : API prefix 설정 문제 해결

Fixes: #31

kym8821 committed:
test: [#31 - ]DesignBoard 통합 테스트 코드 추가 및 테스트 데이터 생성 기능 개선

테스트 데이터 생성 기능에서 테마 및 디자인 게시판 테스트 데이터를 생성 가능하도록 개선
디자인 게시판 생성, 조회 통합 테스트 코드 작성

Fixes: #31

kym8821 committed:
test: [#31 - ] DesignBoard 수정, 삭제 테스트 작성 및 assertion 보완

DesignBoardDetailDTO에 대한 assertion 시 DB에 있는 값도 확인하도록 개선 및 공통 기능으로 분리
ThemeBoardDataGenerator 삭제
FileManager 전역 모킹 시 인터페이스를 기반으로 하도록 개선

Fixes: #31

kym8821 committed:
chore: [#31 - ] 불필요한 파일 제거

TagManagementFacade 제거

Fixes: #31

kym8821 committed:
test: [#31 - ]테스트 코드에 content tye 추가

ThemeManageControllerTest에서 요청 시 content-type: Application_json 추가

klaqwe1 commented on src/main/java/com/komentum/post/controller/CommentController.java:
기본값을 설정하면, 코드가 간결해질 뿐더러 null 체크를 따로 하지 않아도 돼서 안정성이 증가하겠네요! 좋습니다 :)

klaqwe1 commented on src/main/java/com/komentum/global/utils/S3FileManager.java:
```diff
@@ -36,7 +36,7 @@ public S3FileManager(@Value("${aws.s3.access-key}") String accessKey,
   }
 
-  private String resolveFilePath(String fileName) {
+  public String resolveFilePath(String fileName) {
```
private에서 public으로 바꾸게 된 계기가 궁금합니다

kym8821 replied:
profile image url을 반환하기 위해서는 s3에 저장된 파일 이름이 필요합니다. 따라서 현재 fileName을 post 테이블에 저장하도록 수정한 상태이고, 해당 fileName을 기반으로 profile image url을 반환할 수 있는 기능이 필요했습니다.

이러한 이유로 기존에 private이었던 resolveFilePath를 public으로 전환하고, 해당 기능을 인터페이스에 추가하여 다른 객체에서 fileName을 실제 file url로 반환할 수 있도록 했습니다.

klaqwe1 commented on src/main/java/com/komentum/global/utils/DateUtils.java:
재사용성을 위해 이런 식으로 따로 빼두는 게 아주 좋은 것 같습니다! :)

kym8821 replied:
YYYY-MM-dd 포멧으로 날짜 문자열을 생성해준다고 보면 됩니다

klaqwe1 commented on src/main/java/com/komentum/post/facade/CommentManagementFacade.java:
facade 패턴을 적용하셨는데, 굳이 이런 구조를 채택하신 이유가 궁금합니다

kym8821 replied:
기존에는 Service 레이어가 모든 비즈니스 로직을 처리하고 있었기 때문에 다음과 같은 문제가 발생했습니다.

- DB 접근 이전에 필요한 규칙 검증이나 도메인 로직까지 Service 레이어에서 처리하면서, 해당 로직을 다른 기능에서 재사용하기 어려움

- 여러 Service 클래스가 서로의 기능을 호출하면서 순환 참조 위험이 높아지고, 객체 간 의존관계가 복잡하게 얽혀 관리가 어려움

이를 해결하기 위해 엔티티 단위의 도메인 규칙 및 검증은 Domain Service로 분리하고, 여러 도메인을 조합해야 하는 유즈케이스 중심의 로직은 Facade에서 orchestration하도록 구조를 재정비했습니다.

그 결과, 아래 효과를 얻을 수 있으리라 생각합니다
- Domain Service가 도메인 규칙을 명확하게 담당하게 되어 재사용성이 높아짐
- 복합 시나리오를 Facade에서 단일 진입점으로 관리할 수 있어 의존관계 복잡도가 감소함
- Service 레이어의 역할이 명확해져 코드 구조와 유지보수성이 개선됨

klaqwe1 commented on src/main/java/com/komentum/post/controller/CommentController.java:
```diff
@@ -23,14 +26,15 @@
-      @RequestParam("pageNumber") Integer pageNumber, @RequestParam("pageSize") Integer pageSize) {
-    pageNumber = pageNumber == null ? 0 : pageNumber;
-    pageSize = pageSize == null ? 20 : pageSize;
+      @Valid @ModelAttribute PageableRequestDto pageableRequestDto) {
```
@valid를 사용하니까 책임을 DTO로 위임할 수 있네요.
컨트롤러의 로직을 깔끔하게 관리하고, 일관적인 에러 처리를 할 수 있게 됐네요!
매우 좋은 리팩토링인 것 같습니다!

klaqwe1 commented on src/main/java/com/komentum/post/controller/CommentController.java:
```diff
@@ -39,10 +43,11 @@ public ResponseEntity<CommentResponse> getComment(@PathVariable Long commentId)
+    return ResponseEntity.ok(commentManagementFacade.createCommentOnPost(postId, createDto));
   }
 
   @PutMapping("/comments/{commentId}")
```
create 메서드는 facade 패턴을 적용하였지만,
나머지의 경우엔 적용하지 않은 이유가 궁금합니다.

kym8821 replied:
이 코드는 두 문제에 대해 고민하여 결론을 냈습니다.
- facade를 단일 진입점으로 유지해야하는 문제
- facade가 단순히 호출만 중계하는 역할을 하게 되어 불필요한 계층이 발생하는 문제

현재 도메인 서비스만 사용하는 다른 메서드들은 post entity만을 사용하기에 현재 상황에서 불필요한 코드와 계층만 증가한다고 판단했습니다.
다만, 확장성의 여지가 높은 mvp 개발하는 상황이기에 리뷰자께서 말씀하신 부분도 충분히 일리가 있다고 생각하며 추후 문제 발생 시 facade 패턴을 적용할 예정입니다.

klaqwe1 commented on src/main/java/com/komentum/user/service/UserRetrieveService.java:
```diff
@@ -15,8 +15,12 @@ public UserRetrieveService(UserRepository userRepository) {
   public UserResponseDto getUserByEmail(String email) {
-    User user = userRepository.findById(email)
-        .orElseThrow(() -> new RuntimeException("user not found"));
+    User user = findUserEntity(email);
```
이렇게 작은 단위의 기능도 따로 메서드로 관리하는 게 좋은가요? 궁금해서 물어봅니다

kym8821 replied:
작은 단위이더라도 데이터 검증 및 Entity 반환을 도메인 서비스 레이어에 분리했을 때, 데이터에 대한 정책 충족과 재사용성 측면에서 적절한 분리라고 생각했습니다.

이러한 기능은 애플리케이션 수준에서 공통 조건 검증 및 예외 처리로써 다른 비즈니스 로직 처리 메서드들에서 재사용되는 경우 충분한 역할을 할 것이라 생각합니다.

klaqwe1 commented on src/main/java/com/komentum/post/dto/DesignBoardDto.java:
```diff
@@ -0,0 +1,145 @@
+  @AllArgsConstructor
+  public static class DesignBoardDetailDto {
+
+    @JsonProperty("board_id")
```
자바에선 카멜케이스를 쓰고 JSON에서는 snake_case를 사용할 때 유용한 어노테이션이네요! 하나 배워갑니다

klaqwe1 commented on src/main/java/com/komentum/post/service/PostService.java:
```diff
@@ -19,43 +15,31 @@
   public void deletePost(Long postId) {
-    PostRawData target = getPostById(postId);
-    Post targetPost = target.getPost();
+    Post targetPost = getPostByPostId(postId);
```
delete하게 되면 게시글의 하위 항목인 댓글, prefer 같은 연관관계를 잘 지울 수 있나 궁금합니다

kym8821 replied:
Tag, Prefer, Post는 현재 on_delete=cascade 설정을 해둔 상태입니다.
따라서 Post가 삭제되면 해당 연관 테이블들도 자동으로 삭제됩니다.

klaqwe1 commented on src/main/java/com/komentum/post/repository/PostRepositorySupport.java:
```diff
@@ -0,0 +1,58 @@
+        .build();
+  }
+
+  public List<PostSummary> findPostSummaries(Pageable pageable) {
```
이게 dsl의 힘이군요 대단합니다

klaqwe1 commented on src/main/java/com/komentum/post/repository/TagRepository.java:
```diff
@@ -3,10 +3,20 @@
+  @Query("select t from Tag t where t.tagId in :postIds")
+  List<Tag> findAllByPostIds(@Param("postIds") List<Long> postIds);
+
+  @Query("select t from Tag t " +
```
이 정도의 질의에는 dsl을 쓸 필요가 없는 지 궁금합니다.
그렇다면, query dsl과 jpql 중 어느 걸 사용할 지, 기준이 있을까요?

klaqwe1 commented on src/main/java/com/komentum/post/dto/ThemeBoardDto.java:
```diff
@@ -0,0 +1,149 @@
+    @JsonProperty("created_at")
+    private String createdAt;
+
+    @JsonProperty("profile_image_url")
```
프로필이미지는 User의 속성인데, createDto에 있을 필요는 없다고 생각합니다.

kym8821 replied:
프로필 이미지는 게시글의 미리보기 이미지입니다!

klaqwe1 commented on src/main/java/com/komentum/post/controller/TagController.java:
```diff
@@ -29,13 +27,6 @@ public ResponseEntity<List<TagResponse>> getTagsByTagId(@PathVariable Long postI
         tagService.findAllByPostId(postId).stream().map(TagResponse::from).toList());
   }
 
-  @PostMapping("/{postId}/tags")
```
createTag를 controller layer에서 삭제한 이유가 궁금합니다

kym8821 replied:
tag의 생명주기가 post를 따라가기에 불필요한 기능이라고 생각했고, 그렇기에 제거했습니다.

klaqwe1 commented on src/main/java/com/komentum/post/dto/DesignBoardDto.java:
```diff
@@ -0,0 +1,145 @@
+  @Builder
+  @NoArgsConstructor
+  @AllArgsConstructor
+  public static class DesignBoardPreviewDto {
```
designboardpreview의 역할이 궁금합니다

kym8821 replied:
DesignBoardPreviewDto는 디자인 게시판 목록을 반환할 때 사용하는 응답용 DTO입니다. 
따라서 DesignBoardDetailDto와 비교했을 때 보다 정보를 요약해서 제공합니다.

klaqwe1 replied:
아, 목록 - 상세 간의 차이이군요. 이해했습니다

kym8821 committed:
style: [#31 - ] 게시판의 profile image 변수명을 preview image로 변경

DTO, service, domain, controller에서 사용하는 게시글의 profile_image 변수명을 preview_image로 변경

Fixes: #31
